tools/sof_perf_analyzer.py

```diff
@@ -201,11 +206,40 @@ def process_kmsg_file():
                 # widgets in the same pipeline and pipelines are created sequentially.
                 if widget_name.startswith('pipeline'):
                     ppln_id = line_split[2]
-                next_line = next(f)
+                next_line = next(f, None)
+                if not next_line:
+                    # reach EOF, exit.
+                    break
                 widget_id = next_line.split('|')[0].split(':')[-1].strip()
                 # convert to the same ID format in mtrace
-                widget_id = int('0x' + widget_id[-6:], 16)
+                try:
+                    widget_id = int('0x' + widget_id[-6:], 16)
+                except ValueError:
+                    print(f'no valid widget ID found for {widget_name}, try with bind message')
+                    not_found[widget_name] = ppln_id
+                    continue
                 comp_name[str(Component(ppln_id, widget_id))] = widget_name
+            elif match_obj := re.search(r'bind.+->.+', line):
+                if not_found:
+                    # parse current line to get widget names
+                    widget_names = match_obj.group(0).split()[1::2]
+                    widget_names = [name.split(':')[0] for name in widget_names]
+
+                    # parse next line to get widget IDs
+                    next_line = next(f, None)
+                    if not next_line:
+                        # reach EOF, exit.
+                        break
+                    widget_ids = next_line.split(': ')[-2].split('|')
+                    widget_ids = int('0x' + widget_ids[0][-6:], 16), int(widget_ids[1], 16)
+
+                    widgets = dict(zip(widget_names, widget_ids))
+
+                    # search and set missed comp_name
+                    for widget_name, widget_id in widgets.items():
+                        if ppln_id := not_found.get(widget_name):
```

Comment:
This `not_found` makes the code significantly more complicated for no obvious benefit, please get rid of it. It should be as simple as:

1) Use the `for line` loop to build some `id + name` dictionary. Don't touch `comp_name` yet. When you find a `bind` line, simply give `Creating` lines precedence in the temporary dictionary.

2) AFTER the `for line` loop, use the temporary and complete dictionary to create `comp_name`.

Reply:
@marc-hb Component class have pipeline_id and comp_id, two parts here. 
from `create widget` message we could parse widget(component) name and pipeline id, and from next line we parse comp_id. on next line parsing error, we only have pipeline id and widget name.

from bind message, we parse widget name, and from its next line, we parse comp_id. bind message is complementary information if we failed to parse some 'create message'. 

in both `create widget` and bind message parsing, two lines are involved. we cannot assume it is atomic. because the concurrent nature of the linux kernel.

We are use two types of information to get more Component class here.  there is still probability a component in the topology is missing in our comp_name dict, but the probability is low as we parse more and more information ('create widget', 'bind message', 'unbind message'), but in practice, with 'create widget' and bind message,  the error rate can be think as low as zero.

Reply:
Thanks @aiChaoSONG for the explanation. The length of your explanation demonstrates that the data being parsed is complex for various reasons. It's not possible to review this parsing code for correctness without understanding the data it's trying to parse. So examples are very useful but they are not enough. I'm afraid this is one of these rare cases where it's very hard to make the code "self-explanatory" and where a lot more comments than usual are required in the source code. For sure a lot of what you just wrote should make its way to the source code.

People's assignments come and go and we cannot have only 1.5 people in the team who can understand this code (or any code for that matter)

"“Programs must be written for people to read, and only incidentally for machines to execute.” - this is especially true in Python code which is very slow to execute (but usually much faster to read).

Also, forcing oneself to explain the code (with comments) very often results in finding simplifications; writing comments is a good example of https://en.wikipedia.org/wiki/Rubber_duck_debugging

Reply:
@aiChaoSONG shall we go with itertools.pairwise？